govuk_chat_evaluation/output_guardrails/evaluate.py

```diff
@@ -0,0 +1,149 @@
+            case (False, True):
+                return "false_positive"
+            case (True, False):
+                return "false_negative"
+
+    @property
+    def classification_exact(self) -> str:
+        is_match = self.expected_exact == self.actual_exact
+        expected_is_positive = self.expected_exact.startswith("True")
+
+        if is_match:
+            return "true_positive" if expected_is_positive else "true_negative"
+        else:
+            return "false_positive" if not expected_is_positive else "false_negative"
+
+    def for_csv(self) -> dict[str, Any]:
+        return {
+            **self.model_dump(),
+            "classification_triggered": self.classification_triggered,
+            "classification_exact": self.classification_exact,
+        }
+
+
+class AggregateResults:
+    def __init__(self, evaluation_results: List[EvaluationResult]):
+        self.evaluation_results = evaluation_results
+        counter_triggered = Counter(
+            evaluation_result.classification_triggered
+            for evaluation_result in evaluation_results
+        )
+        counter_exact = Counter(
+            evaluation_result.classification_exact
+            for evaluation_result in evaluation_results
+        )
+        self.true_positive_triggered = counter_triggered.get("true_positive", 0)
+        self.true_negative_triggered = counter_triggered.get("true_negative", 0)
+        self.false_positive_triggered = counter_triggered.get("false_positive", 0)
+        self.false_negative_triggered = counter_triggered.get("false_negative", 0)
+        self.true_positive_exact = counter_exact.get("true_positive", 0)
+        self.true_negative_exact = counter_exact.get("true_negative", 0)
+        self.false_positive_exact = counter_exact.get("false_positive", 0)
+        self.false_negative_exact = counter_exact.get("false_negative", 0)
+
+    @cached_property
+    def _expected_actual_lists_triggered(self) -> tuple[List[int], List[int]]:
+        pairs_list = [
+            (
+                int(evaluation_result.expected_triggered),
+                int(evaluation_result.actual_triggered),
+            )
+            for evaluation_result in self.evaluation_results
+        ]
+        expected, actual = zip(*pairs_list)
+        return list(expected), list(actual)
+
+    def precision_triggered(self) -> float:
+        return precision_score(
+            *self._expected_actual_lists_triggered,
+            zero_division=np.nan,  # type: ignore
+        )
+
+    def recall_triggered(self) -> float:
+        return recall_score(
+            *self._expected_actual_lists_triggered,
+            zero_division=np.nan,  # type: ignore
+        )
+
+    def _calculate_metric(self, numerator: int, denominator: int) -> float:
+        if denominator == 0:
+            return float("nan")
+        return float(numerator) / denominator
+
+    def precision_exact(self) -> float:
+        return self._calculate_metric(
```

Comment:
It is a bit surprising that we have two different techniques in this class to calculate precision and recall. I'm expecting we went with this different technique because it wasn't as easy to create the lists.

Assuming it is a bit of a pain to create the lists and thus use sklearn maybe we want to just use the helper function. 

In doing so I think we should rename from the helper of "_calculate_metric" as that is very non specific (and really is just a division function) to a `_calculate_precision(true_positives, false_positives)` and `_calculate_recall(true_positives, false_negatives)`